### PR TITLE
Move type to file: allow syncing filename and type name if no type in file matches filename

### LINKED ISSUE #12727: Move type to file: allow syncing filename and type name if no type in file matches filename
**Steps to Reproduce**:
current design offers to sync filename or typename only if there is a single toplevel type decl in a file and it doesn't match file name.

A better design here would be to look through all the top level type declarations, and if none of them match the filename, allow sync -ing to file name on all those types.

/cc: thanks @CyrusNajmabadi .

### PR DESCRIPTION
**Pull request by balajikris**
Fixes #12727 

Look through all the top level type declarations in a file and if

1. none of them match the filename, offer "rename type to match file" and "rename file to match type" code actions on each of those types.
2. at least one such type matches file name, do not offer "rename file/type" codeactions on any of those types.

### CODE REVIEW

**Comment by balajikris** [comment]
@dotnet/roslyn-ide : Please review.

**Note to reviewers:** Please review this, alongside of https://github.com/dotnet/roslyn/pull/12812, as the changes are in the same areas and the latter PR improves upon the changes introduced here.

File: src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs [hunk 1]
```diff
@@ -69,21 +50,28 @@ protected virtual SyntaxNode GetNodeToAnalyze(SyntaxNode root, TextSpan span)
...
             {
-                // If there are multiple type declarations in current document. offer, move to new file.
-                // Or if this is a nested type, offer to move to new file.
                 actions.Add(GetCodeAction(state, operationKind: OperationKind.MoveType));
             }
-            else
+
+            // (2) Add rename file and rename type code actions:
+            // case 1: Multiple type decls in file and none match file name.
+            // case 2: Single type declaration in file and names don't match.
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs]: Case 1 and 2 are the same "No type declaration in the file matches the file name."

File: src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs [hunk 2]
```diff
@@ -97,11 +85,33 @@ private ImmutableArray<CodeAction> CreateActions(State state, CancellationToken
...
+            new MoveTypeCodeAction((TService)this, state, operationKind);
+
+        private bool IsNestedType(TTypeDeclarationSyntax typeNode) =>
+            typeNode.Parent is TTypeDeclarationSyntax;
+
+        /// <summary>
+        /// checks if there is a single top level type declaration in a document
+        /// </summary>
+        private bool MultipleTopLevelTypeDeclarationInSourceDocument(SyntaxNode root) =>
+            TopLevelTypeDeclarations(root).Count() > 1;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs]: Count is linear.  You can just do ```.Any```

File: src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs [hunk 3]
```diff
@@ -97,11 +85,33 @@ private ImmutableArray<CodeAction> CreateActions(State state, CancellationToken
...
         {
-            return new MoveTypeCodeAction((TService)this, state, operationKind);
+            var root = state.SemanticDocument.Root;
+            var semanticModel = state.SemanticDocument.SemanticModel;
+
+            return TopLevelTypeDeclarations(root).Any(
+                typeDeclaration => 
+                {
+                    var typeName = semanticModel.GetDeclaredSymbol(typeDeclaration, cancellationToken).Name;
+                    return string.Equals(state.DocumentName, typeName, StringComparison.CurrentCulture);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs]: Doesn't this mean that i fi have "Outer.Inner.cs" and you'll prompt to rename this file to "Outer.cs"?

File: src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs]: Oh, you actually care if it's greater than 1.

File: src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs (hunk 1 above)
Comment (balajikris) [inline src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs]: true, I do not differentiate between them in code, just elucidating in comments.

File: src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs (hunk 3 above)
Comment (balajikris) [inline src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs]: In this PR, yes. But, I've improved this code in the other PR I have open (tagged it in this PR) which handles the Outer.Inner cases much better and that will not prompt in your scenario.

File: src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs]: Interestingly, i find that more confusing :)   can you make the comment state that tehy're both effectively the same thing

**Comment by Pilchie** [comment]
👍, though I prefer the `Skip(1).Any()` instead of `Count() > 1` trick.

**Comment by balajikris** [comment]
@Pilchie : thanks, done.

**Pull request merged**